manifest/v1alpha/slo/validation.go

```diff
@@ -24,6 +33,97 @@ var metadataValidation = validation.New[Metadata](
 	validationV1Alpha.FieldRuleMetadataLabels(func(m Metadata) v1alpha.Labels { return m.Labels }),
 )
 
+var specValidationNonComposite = validation.New[Spec](
+	validation.For(func(s Spec) Indicator { return s.Indicator }).
+		WithName("indicator").
+		Required().
+		Include(indicatorValidation),
+).When(func(s Spec) bool { return !s.HasCompositeObjectives() })
+
+var specValidationComposite = validation.New[Spec](
+	validation.For(func(s Spec) Indicator { return s.Indicator }).
+		WithName("indicator").
+		Rules(
+			validation.Forbidden[Indicator]().WithDetails(
+				"indicator section is forbidden when spec.objectives[0].composite is provided",
+			),
+		).
+		StopOnError(),
+
+	validation.ForPointer(func(s Spec) *Composite { return s.Composite }).
+		WithName("composite").
+		Rules(
+			validation.Forbidden[Composite]().WithDetails(
+				"composite section is forbidden when spec.objectives[0].composite is provided",
+			),
+		).
+		StopOnError(),
+
+	validation.For(func(s Spec) Spec { return s }).
+		Rules(specCompositeObjectiveValidationRule).
+		StopOnError(),
+
+	validation.Transform(func(s Spec) string { return s.Objectives[0].Composite.MaxDelay }, time.ParseDuration).
+		WithName("objectives[0].composite.maxDelay").
+		Rules(
+			validation.DurationPrecision(time.Minute),
+			validation.GreaterThanOrEqualTo(time.Minute),
+		).
+		StopOnError(),
+
+	// Weights of composite objectives must be greater than zero
+	validation.ForEach(func(s Spec) []CompositeObjective { return s.Objectives[0].Composite.Objectives }).
```

Comment:
I don't like that we're skipping levels like that in here :/
While sometimes it might be necessary, I think we can safely drill down without skipping anything here, take a look at a refactor proposal I've made to mitigate that:

```go
	validation.ForEach(func(s Spec) []Objective { return s.Objectives }).
		WithName("objectives").
		Rules(validation.SliceLength[[]Objective](1, 1)).
		IncludeForEach(validation.New[Objective](
			validation.ForPointer(func(o Objective) *CompositeSpec { return o.Composite }).
				WithName("composite").
				Include(validation.New[CompositeSpec](
					validation.Transform(func(c CompositeSpec) string { return c.MaxDelay }, time.ParseDuration).
						WithName("maxDelay").
						Rules(
							validation.DurationPrecision(time.Minute),
							validation.GreaterThanOrEqualTo(time.Minute),
						),
					validation.ForEach(func(c CompositeSpec) []CompositeObjective { return c.Objectives }).
						WithName("objectives").
						IncludeForEach(compositeObjectiveRule),
				)),
		)),
```

Reply:
Rewrote using the proposal.